plugins/inputs/redfish/redfish.go

```diff
@@ -0,0 +1,312 @@
+}
+type Redfish struct {
+	Host              string `toml:"host"`
+	BasicAuthUsername string `toml:"username"`
+	BasicAuthPassword string `toml:"password"`
+	Id                string `toml:"id"`
+	client            http.Client
+	tls.ClientConfig
+	Timeout internal.Duration `toml:"timeout"`
+}
+
+func (r *Redfish) Description() string {
+	return "Read CPU, Fans, Powersupply and Voltage metrics of Dell/HP hardware server through redfish APIs"
+}
+
+var redfishConfig = `
+  ##Server  OOB-IP
+  host = "192.0.0.1"
+
+  ##Username,  Password   for   hardware   server
+  username = "test"
+  password = "test"
+
+  ##Resource  Id   for   redfish   APIs
+  id="System.Embedded.1"
+
+  ##Optional TLS   Config, if not provided insecure skip verifies defaults to true
+  #tls_ca = "/etc/telegraf/ca.pem"
+  #tls_cert = "/etc/telegraf/cert.pem"
+  #tls_key = "/etc/telegraf/key.pem"
+  
+
+  ## Amount   of   time   allowed   to   complete   the   HTTP   request
+  # timeout = "5s"
+`
+
+func (r *Redfish) SampleConfig() string {
+	return redfishConfig
+}
+
+func (r *Redfish) Init() error {
+	tlsCfg, err := r.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+	if (len(r.ClientConfig.TLSCA) == 0) || (len(r.ClientConfig.TLSCert) == 0 && len(r.ClientConfig.TLSKey) == 0) {
+		var insecuretls tls.ClientConfig
+		insecuretls.InsecureSkipVerify = true
+		tlsCfg, err = insecuretls.TLSConfig()
+		if err != nil {
+			return err
+		}
+	}
+	r.client = http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+			Proxy:           http.ProxyFromEnvironment,
+		},
+		Timeout: r.Timeout.Duration,
+	}
+	return nil
+}
+
+func (r *Redfish) Gather(acc telegraf.Accumulator) error {
+	var url []string
+	var payload Payload
+	url = append(url, fmt.Sprint("https://", r.Host, "/redfish/v1/Chassis/", r.Id, "/Thermal"), fmt.Sprint("https://", r.Host, "/redfish/v1/Chassis/", r.Id, "/Power"), fmt.Sprint("https://", r.Host, "/redfish/v1/Systems/", r.Id), fmt.Sprint("https://", r.Host, "/redfish/v1/Chassis/", r.Id, "/"))
+
+	if len(r.Host) == 0 || len(r.BasicAuthUsername) == 0 || len(r.BasicAuthPassword) == 0 {
+		return fmt.Errorf("Did not provide IP or username and password")
+	}
+	if len(r.Id) == 0 {
+		return fmt.Errorf("Did not provide all the ID of the resource")
+	}
+
+	for _, i := range url {
+		req, err := http.NewRequest("GET", i, nil)
+		if err != nil {
+			return err
+		}
+		req.SetBasicAuth(r.BasicAuthUsername, r.BasicAuthPassword)
+		req.Header.Set("Accept", "application/json")
+		req.Header.Set("Content-Type", "application/json")
+		resp, err := r.client.Do(req)
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode == 200 {
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			jsonErr := json.Unmarshal(body, &payload)
+			if jsonErr != nil {
+				return fmt.Errorf("error parsing input: %v", jsonErr)
+			}
+
+		} else {
+			return fmt.Errorf("received status code %d (%s), expected 200",
```

Comment:
This stops processing as soon as it finds a url that doesn't return a 200. This is probably okay, but you might want to return whatever data you can manage, instead of no data.

Reply:
It would be appropriate to stop the process when it does not return 200, in some scenarios for example user might give wrong credentials or ID, the process will stop at 1st API call and also if there is a proxy or any connectivity problem the process will be stopped at 1st API call itself. there might be rare situations where we receive data for one API and not for other APIs.